testing/base/source_gen_test.cpp

```diff
@@ -0,0 +1,173 @@
+}
+
+TEST(SourceGenTest, Ids) {
+  SourceGen gen;
+
+  auto ids = gen.GetShuffledIds(1000);
+  EXPECT_THAT(ids.size(), Eq(1000));
+  for (llvm::StringRef id : ids) {
+    EXPECT_THAT(id, MatchesRegex("[A-Za-z][A-Za-z0-9_]*"));
+  }
+
+  // Check that repeated calls are different in interesting ways, but have the
+  // exact same total bytes.
+  ssize_t ids_size_sum = SumSizes(ids);
+  for ([[maybe_unused]] int i : llvm::seq(1, 10)) {
+    auto ids2 = gen.GetShuffledIds(1000);
+    EXPECT_THAT(ids2, SizeIs(1000));
+    // Should be (at least) a different shuffle of identifiers.
+    EXPECT_THAT(ids2, Not(ContainerEq(ids)));
+    // But the sum of lengths should be identical.
+    EXPECT_THAT(SumSizes(ids2), Eq(ids_size_sum));
+  }
+
+  // Check length constraints have the desired effect.
+  ids = gen.GetShuffledIds(1000, /*min_length=*/10, /*max_length=*/20);
+  for (llvm::StringRef id : ids) {
+    EXPECT_THAT(id.size(), Ge(10)) << "Too short id: " << id;
+    EXPECT_THAT(id.size(), Le(20)) << "Too long id: " << id;
+  }
+
+  // Check that uniform id length results in exact coverage of each possible
+  // length for an easy case.
+  ids = gen.GetShuffledIds(100, /*min_length=*/10, /*max_length=*/19,
+                           /*uniform=*/true);
+  EXPECT_THAT(ids, Contains(SizeIs(10)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(11)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(12)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(13)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(14)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(15)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(16)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(17)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(18)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(19)).Times(10));
+}
+
+auto TestCompile(llvm::StringRef source) -> bool {
+  llvm::vfs::InMemoryFileSystem fs;
+  InstallPaths installation(
+      InstallPaths::MakeForBazelRunfiles(Testing::GetTestExePath()));
+  Driver driver(fs, &installation, llvm::outs(), llvm::errs());
+
+  // Load the prelude into our VFS.
+  //
+  // TODO: Factor this and analogous code in file_test into a Driver helper.
+  auto prelude =
+      Driver::FindPreludeFiles(installation.core_package(), llvm::errs());
+  CARBON_CHECK(!prelude.empty());
+  for (const auto& path : prelude) {
+    llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> file =
+        llvm::MemoryBuffer::getFile(path);
+    CARBON_CHECK(file) << file.getError().message();
+    CARBON_CHECK(fs.addFile(path, /*ModificationTime=*/0, std::move(*file)))
+        << "Duplicate file: " << path;
+  }
+
+  fs.addFile("test.carbon", /*ModificationTime=*/0,
+             llvm::MemoryBuffer::getMemBuffer(source));
+  return driver.RunCommand({"compile", "--phase=check", "test.carbon"}).success;
+}
+
+TEST(SourceGenTest, GenAPIFileDenseDeclsTest) {
+  SourceGen gen;
+
+  std::string source =
+      gen.GenAPIFileDenseDecls(1000, SourceGen::DenseDeclParams{});
```

Comment:
What do you think of commenting the 1000 parameter here, to give a hint about what it is?